config/reader_test.go

```diff
@@ -2,16 +2,137 @@ package config
 
 import (
 	"context"
+	"fmt"
 	"os"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/google/uuid"
+	pb "go.viam.com/api/app/v1"
 	"go.viam.com/test"
 
+	"go.viam.com/rdk/config/testutils"
 	"go.viam.com/rdk/logging"
 )
 
+func TestFromReader(t *testing.T) {
+	const (
+		robotPartID = "forCachingTest"
+		secret      = testutils.FakeCredentialPayLoad
+	)
+	var (
+		logger = logging.NewTestLogger(t)
+		ctx    = context.Background()
+	)
+
+	// clear cache
+	setupClearCache := func(t *testing.T) {
+		t.Helper()
+		clearCache(robotPartID)
+		_, err := readFromCache(robotPartID)
+		test.That(t, os.IsNotExist(err), test.ShouldBeTrue)
+	}
+
+	setupFakeServer := func(t *testing.T) (*testutils.FakeCloudServer, func()) {
+		t.Helper()
+
+		logger := logging.NewTestLogger(t)
+
+		fakeServer, err := testutils.NewFakeCloudServer(context.Background(), logger)
+		test.That(t, err, test.ShouldBeNil)
+		cleanup := func() {
+			test.That(t, fakeServer.Shutdown(), test.ShouldBeNil)
+		}
+
+		return fakeServer, cleanup
+	}
+
+	t.Run("online", func(t *testing.T) {
+		setupClearCache(t)
+
+		fakeServer, cleanup := setupFakeServer(t)
+		defer cleanup()
+
+		cloudResponse := &Cloud{
+			ManagedBy:        "acme",
+			SignalingAddress: "abc",
+			ID:               robotPartID,
+			Secret:           secret,
+			FQDN:             "fqdn",
+			LocalFQDN:        "localFqdn",
+			LocationSecrets:  []LocationSecret{},
+			LocationID:       "the-location",
+			PrimaryOrgID:     "the-primary-org",
+		}
+		certProto := &pb.CertificateResponse{
+			TlsCertificate: "cert",
+			TlsPrivateKey:  "key",
+		}
+
+		cloudConfProto, err := CloudConfigToProto(cloudResponse)
+		test.That(t, err, test.ShouldBeNil)
+		protoConfig := &pb.RobotConfig{Cloud: cloudConfProto}
+		fakeServer.StoreDeviceConfig(robotPartID, protoConfig, certProto)
+
+		appAddress := fmt.Sprintf("http://%s", fakeServer.Addr().String())
+		cfgText := fmt.Sprintf(`{"cloud":{"id":%q,"app_address":%q,"secret":%q}}`, robotPartID, appAddress, secret)
+		gotCfg, err := FromReader(ctx, "", strings.NewReader(cfgText), logger)
+		defer clearCache(robotPartID)
+		test.That(t, err, test.ShouldBeNil)
+
+		expectedCloud := *cloudResponse
+		expectedCloud.AppAddress = appAddress
+		expectedCloud.TLSCertificate = certProto.TlsCertificate
+		expectedCloud.TLSPrivateKey = certProto.TlsPrivateKey
+		expectedCloud.RefreshInterval = time.Duration(10000000000)
+		test.That(t, gotCfg.Cloud, test.ShouldResemble, &expectedCloud)
```

Comment:
https://github.com/viamrobotics/rdk/pull/3686/files#r1525393301

this test fails indicated primary-org-id and location-id are missing unless I add them as described in the linked comment.